wall_e/src/resources/cogs/ban.py

```diff
@@ -0,0 +1,344 @@
+                        f"[{self.config.get_config_value('basic_config', 'BRANCH_NAME')}_mod_channel] "
+                        f"in {self.config.get_config_value('basic_config', 'BRANCH_NAME')} "
+                        "does not exist and I was unable to create it, exiting now...."
+                    )
+                    exit(1)
+                logger.info("[Ban load()] variable "
+                            f"\"mod_channel_id\" is set to \"{mod_channel_id}\"")
+
+    @commands.Cog.listener(name='on_member_join')
+    async def watchdog(self, member: discord.Member):
+        """Watches for users joining the guild and kicks and bans a user if they are banned"""
+
+        if member.id in self.ban_list:
+            logger.info(f"[Ban watchdog()] banned member, {member}, detected. Promply will notify and kick them.")
+
+            e_obj = discord.Embed(title="Ban Notification",
+                                  color=discord.Color.red(),
+                                  timestamp=datetime.datetime.now(pytz.timezone('Canada/Pacific'))
+                                  )
+            e_obj.add_field(name="Notice", value=f"**You are PERMANENTLY BANNED from\n{self.bot.guilds[0]}\n\n"
+                            "You may NOT rejoin the guild!**")
+            e_obj.set_footer(icon_url=self.bot.guilds[0].icon_url, text=self.bot.guilds[0])
+
+            await member.send(embed=e_obj)
+            await member.kick(reason="Not allowed back on server.")
+
+    @commands.Cog.listener(name='on_member_ban')
+    async def intercept(self, guild: discord.Guild, member: Union[discord.User, discord.Member]):
+        """Watches for a guild ban. The guild ban is undone and the user is banned via this ban system"""
+
+        # need to read the audit log to grab mod, date, and reason
+        logger.info(f"[Ban intercept()] guild ban detected and intercepted for user='{member}'")
+        try:
+            audit_ban = await self.bot.guilds[0].audit_logs(action=discord.AuditLogAction.ban).flatten()
+        except Exception as e:
+            logger.info(f'error while fetching ban data: {e}')
+            await self.mod_channel.send(f"Encountered following error while intercepting a ban: {e}\n" +
+                                        "**Most likely need view audit log perms.**")
+            return
+        audit_ban = audit_ban[0]
+        logger.info(f"[Ban intercept()] audit log data retrieved for intercepted ban: {audit_ban}")
+
+        # name, id, mod, mod id, date, reason
+        username = member.name + '#' + member.discriminator
+        mod = audit_ban.user.name + '#' + audit_ban.user.discriminator
+        mod_id = audit_ban.user.id
+        date = audit_ban.created_at
+        reason = audit_ban.reason if audit_ban.reason else 'No Reason Given!'
+
+        # update ban_list and db
+        self.ban_list.append(member.id)
+        await BannedUsers.insert_ban(username, member.id)
```

Comment:
why do you need both the username and the user id? why not just store the user id and then get the username at runtime when you need it again?

Reply:
you cant access users info when they're not a guild member so thats why its stored